File: simulation/generate_all_figures.py
"""
Generate All Figures for the Paper

This script runs all simulations and generates publication-quality figures.
Run from repo root: python -m simulation.generate_all_figures

Usage:
    python -m simulation.generate_all_figures
"""

import sys
import os
import numpy as np
import matplotlib
matplotlib.use('Agg')  # Non-interactive backend
import matplotlib.pyplot as plt

# Ensure project root is on path (for erh_core when run as -m simulation.generate_all_figures)
script_dir = os.path.dirname(os.path.abspath(__file__))
project_root = os.path.dirname(script_dir)
if project_root not in sys.path:
    sys.path.insert(0, project_root)

from erh_core.core.action_space import generate_world
from erh_core.core.judgement_system import BiasedJudge, NoisyJudge, ConservativeJudge, RadicalJudge, batch_evaluate
from erh_core.core.ethical_primes import select_ethical_primes, compute_Pi_and_error, analyze_error_growth, compare_error_distributions
from simulation.analysis.zeta_function import (
    build_m_sequence,
    compute_spectrum,
    find_approximate_zeros,
    analyze_spectrum_peaks,
    fit_error_to_zeta_critical_line,
    detect_zeros,
    detect_poles,
)
from simulation.analysis.statistics import compare_judges, generate_report
from simulation.visualization.plots import (
    setup_paper_style,
    plot_Pi_B_E,
    plot_error_growth,
    plot_multi_judge_errors,
    plot_spectrum,
    plot_zero_distribution,
    plot_complexity_distribution,
    plot_judge_comparison,
    plot_critical_bound,
    plot_phase_transition_diagram,
    plot_ethical_primes_map,
)

# Setup
setup_paper_style()
np.random.seed(42)

print("=" * 70)
print("ETHICAL RIEMANN HYPOTHESIS - FIGURE GENERATION")
print("=" * 70)
print()

# Ensure output directory exists (relative to script directory)
output_dir = os.path.join(script_dir, 'output', 'figures')
os.makedirs(output_dir, exist_ok=True)

print("[1/10] Generating action space (N=2000)...")
actions = generate_world(
    num_actions=2000,
    complexity_dist='zipf',
    complexity_range=(1, 100),
    moral_ambiguity_factor=0.3,
    random_seed=42
)
print(f"      Generated {len(actions)} actions")

print("\n[2/10] Creating judgment systems...")
judges = {
    'Biased': BiasedJudge(bias_strength=0.2, noise_scale=0.1),
    'Noisy': NoisyJudge(noise_scale=0.3),
    'Conservative': ConservativeJudge(threshold=0.5),
    'Radical': RadicalJudge(amplification=1.5)
}
try:
    from simulation.quantum import LocalQuantumJudge
    from simulation.core.judgement_system import QuantumJudge
    oracle = LocalQuantumJudge(shots=512, seed=42)
    judges['Quantum'] = QuantumJudge(quantum_oracle=oracle)
    print("      Added Quantum Judge (qiskit or NumPy fallback)")
except Exception as e:
    print(f"      Quantum Judge skipped: {e}")
print(f"      Created {len(judges)} judges")

print("\n[3/10] Evaluating all judges...")
results = batch_evaluate(actions, judges, tau=0.3)
print("      Evaluation complete")

# Print basic statistics
print("\n" + "=" * 70)
print("PRELIMINARY STATISTICS")
print("=" * 70)
comparison = compare_judges(results, X_max=100)
for name, metrics in comparison.items():
    if 'error' in metrics:
        print(f"\n{name}: ERROR - {metrics['error']}")
        continue
    print(f"\n{name}:")
    alpha = metrics['estimated_exponent']
    within_bound = alpha <= 0.5 + 0.15
    print(f"  Mistake rate: {metrics['mistake_rate']:.3f}")
    print(f"  Num primes: {metrics['num_primes']}")
    print(f"  Estimated exponent α: {alpha:.3f}")
    print(
        "  Within ERH-style bound (α ≲ 0.5)? "
        f"{'Yes' if within_bound else 'No'}"
    )
    # analysis['erh_satisfied'] still encodes \"close to α ≈ 0.5\"
    print(
        "  Close to ERH-style target (α ≈ 0.5)? "
        f"{'Yes [target-like]' if metrics['erh_satisfied'] else 'No'}"
    )
    print(f"  Growth rate: {metrics['growth_rate']}")

print("\n" + "=" * 70)
print("GENERATING FIGURES")
print("=" * 70)

# Figure 1: Basic Π(x), B(x), E(x) for Biased Judge
print("\n[4/11] Figure 1: Π(x), B(x), E(x) curves (Biased Judge)...")
primes_biased = select_ethical_primes(results['Biased'], importance_quantile=0.9)
Pi_x, B_x, E_x, x_vals = compute_Pi_and_error(primes_biased, X_max=100, baseline='prime_theorem')
plot_Pi_B_E(
    x_vals, Pi_x, B_x, E_x,
    title="Ethical Prime Distribution (Biased Judge)",
    save_path=os.path.join(output_dir, 'paper_fig1_pi_b_e.pdf'),
    show=False
)
print(f"      Saved: {os.path.join(output_dir, 'paper_fig1_pi_b_e.pdf')}")

# Figure 2: Error growth analysis
print("\n[5/11] Figure 2: Error growth in log-log scale...")
analysis = analyze_error_growth(E_x, x_vals)
plot_error_growth(
    x_vals, E_x, analysis,
    title="Error Growth Analysis (Biased Judge)",
    save_path=os.path.join(output_dir, 'paper_fig2_error_growth.pdf'),
    show=False
)
print(f"      Saved: {os.path.join(output_dir, 'paper_fig2_error_growth.pdf')}")

# Figure 3: Multi-judge comparison
print("\n[6/11] Figure 3: Multi-judge error comparison...")
error_comparison = compare_error_distributions(results, X_max=100)
plot_multi_judge_errors(
    error_comparison,
    save_path=os.path.join(output_dir, 'paper_fig3_judge_comparison.pdf'),
    show=False
)
print(f"      Saved: {os.path.join(output_dir, 'paper_fig3_judge_comparison.pdf')}")

# Figure 4: Judge metric comparison (bar chart)
print("\n[7/11] Figure 4: Judge exponent comparison...")
plot_judge_comparison(
    comparison,
    metric='estimated_exponent',
    title='Estimated Growth Exponent α by Judge Type',
    save_path=os.path.join(output_dir, 'paper_fig4_exponent_comparison.pdf'),
    show=False
)
print(f"      Saved: {os.path.join(output_dir, 'paper_fig4_exponent_comparison.pdf')}")

# Figure 5: Spectrum analysis
print("\n[8/11] Figure 5: Frequency spectrum...")
m = build_m_sequence(primes_biased, X_max=200, mode='count')
freqs, amps = compute_spectrum(m, normalize=True)
peaks = analyze_spectrum_peaks(freqs, amps, num_peaks=5)
plot_spectrum(
    freqs, amps, peaks,
    title="Frequency Spectrum of Ethical Primes (Biased Judge)",
    save_path=os.path.join(output_dir, 'paper_fig5_spectrum.pdf'),
    show=False
)
print(f"      Saved: {os.path.join(output_dir, 'paper_fig5_spectrum.pdf')}")

# Figure 6: Zero distribution
print("\n[9/11] Figure 6: Ethical zeta function zeros...")
zeros = find_approximate_zeros(
    m,
    real_range=(0.3, 0.7),
    imag_range=(0, 30),
    grid_size=80,
    threshold=0.5,
)
plot_zero_distribution(
    zeros,
    title="Ethical Zeta Function Zeros in Complex Plane",
    save_path=os.path.join(output_dir, 'paper_fig6_zeros.pdf'),
    show=False
)
print(f"      Found {len(zeros)} approximate zeros")
print(f"      Saved: {os.path.join(output_dir, 'paper_fig6_zeros.pdf')}")

# Figure 7: Critical Bound (Log-Log |E(x)| vs x with x^1/2 overlay)
print("\n[10/13] Figure 7: Critical Bound (dual agents)...")
plot_critical_bound(
    error_comparison,
    save_path=os.path.join(output_dir, 'paper_fig7_critical_bound.pdf'),
    show=False
)
print(f"      Saved: {os.path.join(output_dir, 'paper_fig7_critical_bound.pdf')}")

# Figure 8: Ethical Primes Map (2D/3D irreducible dilemmas)
print("\n[11/13] Figure 8: Ethical Primes Map...")
plot_ethical_primes_map(
    primes_biased,
    x_attr='c',
    y_attr='w',
    color_attr='delta',
    title='Ethical Primes: Irreducible Dilemmas in Action Space',
    save_path=os.path.join(output_dir, 'paper_fig8_ethical_primes_map.pdf'),
    show=False
)
print(f"      Saved: {os.path.join(output_dir, 'paper_fig8_ethical_primes_map.pdf')}")

# Figure 9: Zeta zeros/poles analysis
print("\n[12/13] Figure 9: Zeta zeros and poles...")
zeta_fit = fit_error_to_zeta_critical_line(E_x, x_vals)
zeros_x = detect_zeros(E_x, x_vals, threshold=0.15)
poles_x = detect_poles(E_x, x_vals, spike_factor=2.5)
if "error" not in zeta_fit:
    print(f"      Zeta fit: α={zeta_fit.get('fitted_exponent', 0):.3f}, ERH satisfied={zeta_fit.get('erh_satisfied', False)}")
print(f"      Zeros (E≈0): {len(zeros_x)}; Poles (error spikes): {len(poles_x)}")

# Figure 10: Complexity distribution
print("\n[13/13] Figure 10: Complexity distribution...")
plot_complexity_distribution(
    actions,
    title="Action Complexity Distribution (Zipf)",
    save_path=os.path.join(output_dir, 'paper_fig10_complexity_dist.pdf'),
    show=False
)
print(f"      Saved: {os.path.join(output_dir, 'paper_fig10_complexity_dist.pdf')}")

# Figure 11: Phase Transition (if quantum phase script run)
phase_transition_path = os.path.join(script_dir, 'output', 'phase_transition_diagram.png')
if os.path.exists(phase_transition_path):
    print("\n      Phase transition diagram found (run scripts/run_quantum_phase_transition.py --save-plot for full pipeline)")

# Figure 12: Π(x), B(x), E(x) for Quantum Judge
if 'Quantum' in results:
    print("\n[14/14] Figure 12: Π(x), B(x), E(x) curves (Quantum Judge)...")
    primes_quantum = select_ethical_primes(results['Quantum'], importance_quantile=0.9)
    Pi_x_q, B_x_q, E_x_q, x_vals_q = compute_Pi_and_error(primes_quantum, X_max=100, baseline='prime_theorem')
    plot_Pi_B_E(
        x_vals_q, Pi_x_q, B_x_q, E_x_q,
        title="Ethical Prime Distribution (Quantum Judge)",
        save_path=os.path.join(output_dir, 'paper_fig8_quantum_judge.pdf'),
        show=False
    )
    print(f"      Saved: {os.path.join(output_dir, 'paper_fig12_quantum_judge.pdf')}")
else:
    print("\n[14/14] Figure 12: Skipped (Quantum Judge not available)")

# Generate detailed report
print("\n" + "=" * 70)
print("GENERATING REPORT")
print("=" * 70)
output_base = os.path.join(script_dir, 'output')
os.makedirs(output_base, exist_ok=True)
report_path = os.path.join(output_base, 'judge_comparison_report.md')
report = generate_report(results, output_path=report_path, format='markdown')
print(f"\nReport saved to: {report_path}")
print("\nReport preview:")
print("-" * 70)
# Handle encoding for Windows console
try:
    print(report[:1000])
except UnicodeEncodeError:
    # Fallback: print ASCII-safe version
    safe_report = report[:1000].encode('ascii', 'ignore').decode('ascii')
    print(safe_report)
print("...\n")

# Save numerical results
print("\n" + "=" * 70)
print("SUMMARY")
print("=" * 70)
print("\nAll figures generated successfully!")
print("\nGenerated files:")
print(f"  - {os.path.join(output_dir, 'paper_fig1_pi_b_e.pdf')}")
print(f"  - {os.path.join(output_dir, 'paper_fig2_error_growth.pdf')}")
print(f"  - {os.path.join(output_dir, 'paper_fig3_judge_comparison.pdf')}")
print(f"  - {os.path.join(output_dir, 'paper_fig4_exponent_comparison.pdf')}")
print(f"  - {os.path.join(output_dir, 'paper_fig5_spectrum.pdf')}")
print(f"  - {os.path.join(output_dir, 'paper_fig6_zeros.pdf')}")
print(f"  - {os.path.join(output_dir, 'paper_fig7_critical_bound.pdf')}")
print(f"  - {os.path.join(output_dir, 'paper_fig8_ethical_primes_map.pdf')}")
print(f"  - {os.path.join(output_dir, 'paper_fig10_complexity_dist.pdf')}")
if 'Quantum' in results:
    print(f"  - {os.path.join(output_dir, 'paper_fig12_quantum_judge.pdf')}")
print(f"  - {report_path}")
print("\nNext steps:")
print("  1. Review generated figures")
print("  2. Copy figures to ../figures/ for LaTeX inclusion")
print("  3. Update ethical_riemann_hypothesis.tex with results")
print("  4. Compile LaTeX paper: pdflatex ethical_riemann_hypothesis.tex")
print("\n" + "=" * 70)

# Save results to a summary file for LaTeX integration
results_summary_path = os.path.join(output_base, 'results_summary.txt')
with open(results_summary_path, 'w') as f:
    f.write("ETHICAL RIEMANN HYPOTHESIS - NUMERICAL RESULTS\n")
    f.write("=" * 70 + "\n\n")
    
    for name, metrics in comparison.items():
        if 'error' in metrics:
            continue
        
        f.write(f"{name} Judge:\n")
        f.write(f"  Total actions: {metrics['num_actions']}\n")
        f.write(f"  Ethical primes: {metrics['num_primes']}\n")
        f.write(f"  Mistake rate: {metrics['mistake_rate']:.3f}\n")
        f.write(f"  MAE: {metrics['mae']:.3f}\n")
        f.write(f"  RMSE: {metrics['rmse']:.3f}\n")
        alpha = metrics['estimated_exponent']
        within_bound = alpha <= 0.5 + 0.15
        f.write(f"  Estimated exponent: {alpha:.3f}\n")
        f.write(
            "  Within ERH-style bound (α ≲ 0.5)?: "
            f"{'Yes' if within_bound else 'No'}\n"
        )
        f.write(f"  Growth rate: {metrics['growth_rate']}\n")
        f.write(f"  R^2 (fit quality): {metrics['r_squared']:.3f}\n")
        f.write("\n")

print(f"\nNumerical results saved to: {results_summary_path}")

# Write LaTeX table rows for tab:comparison (Judge | Mistake Rate | MAE | F1 | α | ERH | Interpretation)
table_path = os.path.join(output_base, 'figures', 'comparison_table_rows.tex')
os.makedirs(os.path.dirname(table_path), exist_ok=True)
with open(table_path, 'w') as f:
    for name in ('Biased', 'Noisy', 'Conservative', 'Radical'):
        if name not in comparison or 'error' in comparison[name]:
            f.write(f"{name} & [TBD] & [TBD] & [TBD] & [TBD] & [YES/NO] & [TBD] \\\\\n")
            continue
        m = comparison[name]
        mr = m.get('mistake_rate', 0)
        mae = m.get('mae', 0)
        f1 = m.get('f1_score')
        f1_str = f"{f1:.3f}" if f1 is not None else "N/A"
        alpha = m.get('estimated_exponent', 0.5)
        erh = "Yes" if m.get('erh_satisfied', False) else "No"
        interp = "ERH OK" if m.get('erh_satisfied') else "Violates bound"
        f.write(f"{name} & {mr:.3f} & {mae:.3f} & {f1_str} & {alpha:.3f} & {erh} & {interp} \\\\\n")
    if 'Quantum' in comparison and 'error' not in comparison['Quantum']:
        m = comparison['Quantum']
        mr = m.get('mistake_rate', 0)
        mae = m.get('mae', 0)
        f1 = m.get('f1_score')
        f1_str = f"{f1:.3f}" if f1 is not None else "N/A"
        alpha = m.get('estimated_exponent', 0.5)
        erh = "Yes" if m.get('erh_satisfied', False) else "No"
        interp = "ERH OK" if m.get('erh_satisfied') else "Violates bound"
        f.write(f"Quantum & {mr:.3f} & {mae:.3f} & {f1_str} & {alpha:.3f} & {erh} & {interp} \\\\\n")
print(f"LaTeX table rows saved to: {table_path}")
print("\nDone!")

File: julia/scripts/generate_all_figures.jl
#!/usr/bin/env julia
"""
Generate All Figures for the Paper (Julia / CairoMakie)

Produces publication-quality figures equivalent to simulation/generate_all_figures.py.
Output goes to figures/ in the repo root (same paths the Python script writes to).

Usage:
    julia --project=julia julia/scripts/generate_all_figures.jl [--output-dir DIR]
    julia --project=julia julia/scripts/generate_all_figures.jl --smoke
"""

push!(LOAD_PATH, joinpath(@__DIR__, "..", "src"))
using ERH
using CairoMakie
using Random
using Statistics
using JSON3

# ---------------------------------------------------------------------------
# Argument parsing
# ---------------------------------------------------------------------------
function parse_args(argv=ARGS)
    args = Dict{String,Any}(
        "output_dir"  => nothing,
        "smoke"       => false,
        "num_actions" => 2000,
        "X_max"       => 100,
    )
    i = 1
    while i <= length(argv)
        a = argv[i]
        if a == "--smoke"
            args["smoke"] = true
        elseif a == "--output-dir" && i < length(argv)
            i += 1; args["output_dir"] = argv[i]
        elseif a == "--num-actions" && i < length(argv)
            i += 1; args["num_actions"] = parse(Int, argv[i])
        end
        i += 1
    end
    return args
end

# ---------------------------------------------------------------------------
# CairoMakie paper style (mirrors setup_paper_style)
# ---------------------------------------------------------------------------
function setup_paper_style!()
    set_theme!(
        fontsize    = 12,
        figure_padding = 10,
        Axis = (
            spinewidth  = 1,
            xgridvisible = true,
            ygridvisible = true,
            xgridstyle  = :dash,
            ygridstyle  = :dash,
        ),
    )
end

# ---------------------------------------------------------------------------
# Simulation helpers
# ---------------------------------------------------------------------------
function generate_world(; num_actions::Int, complexity_dist::String="zipf", seed::Int=42)
    rng = MersenneTwister(seed)
    [(
        id=i,
        c=if complexity_dist == "zipf"
            clamp(round(Int, 100 / (rand(rng) * 99 + 1)), 1, 100)
        elseif complexity_dist == "uniform"
            rand(rng, 1:100)
        else
            clamp(round(Int, 100 * rand(rng)^2), 1, 100)
        end,
        w=rand(rng),
        v=rand(rng),
        mistake_flag=false,
    ) for i in 1:num_actions]
end

function evaluate_judge!(actions; bias_strength, noise_scale, tau, seed)
    rng = MersenneTwister(seed)
    [(merge(a, (mistake_flag=(abs(clamp(a.v + bias_strength*(0.5-a.v) + randn(rng)*noise_scale, 0.0, 1.0) - a.v) > tau),))) for a in actions]
end

function select_primes(actions; q=0.9)
    thresh = quantile([a.w for a in actions], q)
    filter(a -> a.w >= thresh, actions)
end

function compute_Pi_and_error(primes; X_max=100)
    xs   = collect(1:X_max)
    Pi_x = [count(p -> p.c <= x, primes)        for x in xs]
    B_x  = [x <= 1 ? 1.0 : x / log(x)          for x in xs]
    E_x  = Pi_x .- B_x
    xs, Pi_x, B_x, E_x
end

function fit_alpha(E_x, x_vals)
    valid = [(log(x), log(abs(e) + 1e-9)) for (x, e) in zip(x_vals, E_x) if x > 1 && abs(e) > 1e-9]
    length(valid) < 5 && return 0.5, 0.0
    lx = [v[1] for v in valid]
    ly = [v[2] for v in valid]
    mx, my = mean(lx), mean(ly)
    alpha  = sum((lx .- mx) .* (ly .- my)) / (sum((lx .- mx).^2) + 1e-12)
    ss_res = sum((ly .- (my .+ alpha .* (lx .- mx))).^2)
    ss_tot = sum((ly .- my).^2) + 1e-12
    r2     = 1.0 - ss_res / ss_tot
    alpha, r2
end

function spectrum(primes; X_max=200)
    m    = zeros(Float64, X_max)
    for p in primes
        p.c <= X_max && (m[p.c] += 1)
    end
    n    = length(m)
    F    = abs.(fft_manual(m))
    freqs = (0:(n÷2)) ./ n
    amps  = F[1:(n÷2 + 1)] ./ sum(F[1:(n÷2 + 1)] .+ 1e-12)
    freqs, amps
end

function fft_manual(x::Vector{Float64})
    # Simple DFT (for spectrum visualization; n is small)
    n = length(x)
    [sum(x[k+1] * exp(-2π * im * j * k / n) for k in 0:n-1) for j in 0:n-1]
end

# ---------------------------------------------------------------------------
# Judge configs
# ---------------------------------------------------------------------------
const JUDGES = [
    ("Biased",       (bias_strength=0.2,  noise_scale=0.1, tau=0.3, seed=100)),
    ("Noisy",        (bias_strength=0.0,  noise_scale=0.3, tau=0.3, seed=200)),
    ("Conservative", (bias_strength=0.1,  noise_scale=0.05, tau=0.15, seed=300)),
    ("Radical",      (bias_strength=0.45, noise_scale=0.2,  tau=0.3, seed=400)),
]

# ---------------------------------------------------------------------------
# Figure helpers
# ---------------------------------------------------------------------------
function save_fig(fig, path; dpi=150)
    mkpath(dirname(path))
    save(path, fig; px_per_unit=dpi/96)
    println("      Saved: $path")
end

# ---------------------------------------------------------------------------
# Figure generators
# ---------------------------------------------------------------------------
function fig1_pi_b_e(x_vals, Pi_x, B_x, E_x, path)
    fig = Figure(size=(800, 500))
    ax  = Axis(fig[1,1]; xlabel="x", ylabel="Count / Estimate",
               title="Ethical Prime Distribution (Biased Judge)")
    lines!(ax, x_vals, Float64.(Pi_x); label="Π(x) ethical primes", color=:blue)
    lines!(ax, x_vals, B_x;  label="B(x) baseline (x/ln x)", color=:orange, linestyle=:dash)
    lines!(ax, x_vals, E_x;  label="E(x) = Π(x) - B(x)",    color=:red,    linestyle=:dot)
    axislegend(ax; position=:lt)
    save_fig(fig, path)
end

function fig2_error_growth(x_vals, E_x, alpha, path)
    pos = [(x, abs(e)) for (x, e) in zip(x_vals, E_x) if abs(e) > 1e-9]
    lx  = log.(first.(pos))
    ly  = log.(last.(pos))
    fig = Figure(size=(800, 500))
    ax  = Axis(fig[1,1]; xlabel="log(x)", ylabel="log|E(x)|",
               title="Error Growth Analysis (Biased Judge) — α ≈ $(round(alpha; digits=3))")
    scatter!(ax, lx, ly; color=:blue, markersize=4, label="|E(x)|")
    # Regression line
    if length(lx) > 1
        xfit = range(minimum(lx), maximum(lx); length=50)
        mx, my = mean(lx), mean(ly)
        yfit = my .+ alpha .* (xfit .- mx)
        lines!(ax, collect(xfit), yfit; color=:red, linestyle=:dash, label="fit α=$(round(alpha;digits=3))")
    end
    axislegend(ax; position=:lt)
    save_fig(fig, path)
end

function fig3_multi_judge(judge_results, path)
    fig = Figure(size=(900, 500))
    ax  = Axis(fig[1,1]; xlabel="x", ylabel="|E(x)|",
               title="Multi-Judge Error Comparison")
    colors = [:blue, :orange, :green, :purple, :red]
    for (k, (name, x_vals, _, _, E_x)) in enumerate(judge_results)
        lines!(ax, x_vals, abs.(E_x); label=name, color=colors[mod1(k, length(colors))])
    end
    axislegend(ax; position=:rt)
    save_fig(fig, path)
end

function fig4_exponent_comparison(judge_metrics, path)
    names  = [m[1] for m in judge_metrics]
    alphas = [m[2] for m in judge_metrics]
    fig = Figure(size=(800, 500))
    ax  = Axis(fig[1,1];
        xlabel = "Judge Type",
        ylabel = "Estimated Growth Exponent α",
        title  = "Estimated Growth Exponent α by Judge Type",
        xticks = (1:length(names), names),
    )
    barplot!(ax, 1:length(names), alphas; color=:steelblue)
    hlines!(ax, [0.5]; color=:red, linestyle=:dash, label="ERH bound α=0.5")
    axislegend(ax; position=:rt)
    save_fig(fig, path)
end

function fig5_spectrum(freqs, amps, path)
    top5 = sortperm(amps; rev=true)[1:min(5, length(amps))]
    fig = Figure(size=(800, 500))
    ax  = Axis(fig[1,1]; xlabel="Frequency", ylabel="Normalised Amplitude",
               title="Frequency Spectrum of Ethical Primes (Biased Judge)")
    lines!(ax, freqs, amps; color=:blue)
    scatter!(ax, freqs[top5], amps[top5]; color=:red, markersize=8, label="Peaks")
    axislegend(ax; position=:rt)
    save_fig(fig, path)
end

function fig6_zeros(E_x, x_vals, path)
    # Detect approximate zeros: sign changes in E(x)
    zeros_x = [x_vals[i] for i in 2:length(E_x) if sign(E_x[i]) != sign(E_x[i-1])]
    fig = Figure(size=(700, 700))
    ax  = Axis(fig[1,1]; xlabel="Re(s) proxy", ylabel="Im(s) proxy",
               title="Ethical Zeta Function Zeros in Complex Plane")
    scatter!(ax, fill(0.5, length(zeros_x)), Float64.(zeros_x);
             color=:blue, markersize=8, label="Approximate zeros")
    vlines!(ax, [0.5]; color=:red, linestyle=:dash, label="Critical line Re(s)=0.5")
    axislegend(ax; position=:rt)
    println("      Found $(length(zeros_x)) approximate zeros")
    save_fig(fig, path)
end

function fig7_critical_bound(judge_results, path)
    fig = Figure(size=(900, 500))
    ax  = Axis(fig[1,1]; xlabel="x", ylabel="|E(x)|",
               title="Critical Bound: |E(x)| vs √x",
               xscale=log10, yscale=log10)
    colors = [:blue, :orange, :green, :purple]
    for (k, (name, x_vals, _, _, E_x)) in enumerate(judge_results)
        abse = [max(abs(e), 1e-9) for e in E_x]
        lines!(ax, Float64.(x_vals), abse; label=name, color=colors[mod1(k, length(colors))])
    end
    xs = range(1, maximum(judge_results[1][2]); length=100)
    lines!(ax, xs, sqrt.(xs); color=:black, linestyle=:dash, linewidth=2, label="√x bound")
    axislegend(ax; position=:lt)
    save_fig(fig, path)
end

function fig8_primes_map(primes, path)
    cs = [Float64(p.c) for p in primes]
    ws = [Float64(p.w) for p in primes]
    ds = [Float64(p.v) for p in primes]  # use v as delta proxy
    fig = Figure(size=(800, 600))
    ax  = Axis(fig[1,1]; xlabel="Complexity c", ylabel="Moral weight w",
               title="Ethical Primes: Irreducible Dilemmas in Action Space")
    sc  = scatter!(ax, cs, ws; color=ds, colormap=:viridis, markersize=6)
    Colorbar(fig[1,2], sc; label="Ground-truth value (δ proxy)")
    save_fig(fig, path)
end

function fig10_complexity_dist(actions, path)
    cs = [a.c for a in actions]
    fig = Figure(size=(800, 500))
    ax  = Axis(fig[1,1]; xlabel="Complexity c", ylabel="Count",
               title="Action Complexity Distribution (Zipf)")
    hist!(ax, Float64.(cs); bins=50, color=:steelblue)
    save_fig(fig, path)
end

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
function main()
    args = parse_args()

    if args["smoke"]
        println("[smoke] generate_all_figures: minimal run (50 actions, X_max=20)")
        args["num_actions"] = 50
        args["X_max"]       = 20
    end

    project_root = joinpath(@__DIR__, "..", "..")
    output_dir   = args["output_dir"] !== nothing ? args["output_dir"] : joinpath(project_root, "figures")
    mkpath(output_dir)

    setup_paper_style!()
    Random.seed!(42)

    n_actions = args["num_actions"]
    X_max     = args["X_max"]

    println("=" ^ 70)
    println("ETHICAL RIEMANN HYPOTHESIS — FIGURE GENERATION (Julia / CairoMakie)")
    println("=" ^ 70)

    println("\n[1/10] Generating action space (N=$n_actions)...")
    actions = generate_world(num_actions=n_actions, complexity_dist="zipf", seed=42)
    println("       Generated $(length(actions)) actions")

    println("\n[2/10] Evaluating judges...")
    judge_results = []   # (name, x_vals, Pi_x, B_x, E_x)
    judge_metrics = []   # (name, alpha, r2)
    for (name, params) in JUDGES
        acts    = evaluate_judge!(actions; params...)
        primes  = select_primes(acts; q=0.9)
        xs, Pi_x, B_x, E_x = compute_Pi_and_error(primes; X_max=X_max)
        alpha, r2 = fit_alpha(E_x, xs)
        push!(judge_results, (name, xs, Pi_x, B_x, E_x))
        push!(judge_metrics, (name, alpha, r2))
        erh = 0.4 <= alpha <= 0.65 ? "Yes" : "No"
        println("  $name: α=$(round(alpha; digits=3)) R²=$(round(r2; digits=3)) ERH=$erh")
    end

    biased_idx = findfirst(r -> r[1] == "Biased", judge_results)
    biased_res = judge_results[biased_idx]
    x_vals_b, Pi_x_b, B_x_b, E_x_b = biased_res[2], biased_res[3], biased_res[4], biased_res[5]
    alpha_b = judge_metrics[biased_idx][2]

    println("\n[3/10] Selecting Biased primes for spectrum/zeros...")
    acts_b  = evaluate_judge!(actions; JUDGES[1][2]...)
    primes_b = select_primes(acts_b; q=0.9)

    println("\n[4/10] Figure 1: Π(x), B(x), E(x)...")
    fig1_pi_b_e(x_vals_b, Pi_x_b, B_x_b, E_x_b,
        joinpath(output_dir, "paper_fig1_pi_b_e.pdf"))

    println("\n[5/10] Figure 2: Error growth log-log...")
    fig2_error_growth(x_vals_b, E_x_b, alpha_b,
        joinpath(output_dir, "paper_fig2_error_growth.pdf"))

    println("\n[6/10] Figure 3: Multi-judge error comparison...")
    fig3_multi_judge(judge_results,
        joinpath(output_dir, "paper_fig3_judge_comparison.pdf"))

    println("\n[7/10] Figure 4: Exponent comparison bar chart...")
    fig4_exponent_comparison(judge_metrics,
        joinpath(output_dir, "paper_fig4_exponent_comparison.pdf"))

    println("\n[8/10] Figure 5: Frequency spectrum...")
    freqs, amps = spectrum(primes_b; X_max=min(200, n_actions))
    fig5_spectrum(freqs, amps,
        joinpath(output_dir, "paper_fig5_spectrum.pdf"))

    println("\n[9/10] Figure 6: Zeta zeros...")
    fig6_zeros(E_x_b, x_vals_b,
        joinpath(output_dir, "paper_fig6_zeros.pdf"))

    println("\n[9b/10] Figure 7: Critical bound (log-log)...")
    fig7_critical_bound(judge_results,
        joinpath(output_dir, "paper_fig7_critical_bound.pdf"))

    println("\n[9c/10] Figure 8: Ethical primes map...")
    fig8_primes_map(primes_b,
        joinpath(output_dir, "paper_fig8_ethical_primes_map.pdf"))

    println("\n[10/10] Figure 10: Complexity distribution...")
    fig10_complexity_dist(actions,
        joinpath(output_dir, "paper_fig10_complexity_dist.pdf"))

    # Summary table (LaTeX rows)
    table_path = joinpath(output_dir, "comparison_table_rows.tex")
    open(table_path, "w") do io
        rows = String[]
        for (name, alpha, r2) in judge_metrics
            erh = 0.4 <= alpha <= 0.65 ? "Yes" : "No"
            interp = erh == "Yes" ? "ERH OK" : "Violates bound"
            push!(rows, "$name & [TBD] & [TBD] & N/A & $(round(alpha;digits=3)) & $erh & $interp")
        end
        write(io, join(rows, " \\\\\n") * "\n")
    end
    println("\nLaTeX table rows saved to: $table_path")

    # Numerical summary
    summary_path = joinpath(output_dir, "..", "simulation", "output", "results_summary.txt")
    mkpath(dirname(summary_path))
    open(summary_path, "w") do io
        println(io, "ETHICAL RIEMANN HYPOTHESIS — NUMERICAL RESULTS (Julia)")
        println(io, "=" ^ 70)
        for (name, alpha, r2) in judge_metrics
            erh = 0.4 <= alpha <= 0.65 ? "Yes" : "No"
            println(io, "\n$name Judge:")
            println(io, "  Estimated exponent: $(round(alpha; digits=3))")
            println(io, "  R² (fit quality):   $(round(r2; digits=3))")
            println(io, "  Within ERH bound:   $erh")
        end
    end
    println("Numerical results saved to: $summary_path")

    println("\n" * "=" ^ 70)
    println("All figures generated successfully!")
    println("=" ^ 70)
end

main()
